File: backend/src/main/java/com/app/smartTaskManager/models/Task.java
package com.app.smartTaskManager.models;

import java.time.LocalDateTime;
import java.util.List;

import jakarta.persistence.*;
import lombok.*;


@Entity
@Data
@Table(name = "tasks")
public class Task {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @Column(nullable = false)
    private String title;

    private String description; 
    
    private boolean completed = false;

    private LocalDateTime createdAt = LocalDateTime.now();

    @Enumerated(EnumType.STRING)
    private Priority priority = Priority.MEDIUM;
    public enum Priority {
        LOW, MEDIUM, HIGH
    }

    private LocalDateTime dueDate;
    
    // USER Relation
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    // SubTasks Relation
    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "task_id")
    private List<SubTask> subtasks;

    // Tags Relation
    @ManyToMany(cascade = {CascadeType.PERSIST, CascadeType.MERGE})
    @JoinTable(name = "task_tags",
            joinColumns = @JoinColumn(name = "task_id"),
            inverseJoinColumns = @JoinColumn(name = "tag_id"))
    private java.util.Set<Tag> tags = new java.util.HashSet<>();


}

File: frontend/lib/models/task.dart
import 'sub_task.dart';

class Task {
  final int? id;
  final String title;
  final String? description;
  final bool completed;
  final DateTime? createdAt;
  final String priority;
  final DateTime? dueDate;
  final List<SubTask> subtasks;
  final List<String> tags;

  Task({
    this.id,
    required this.title,
    this.description,
    this.completed = false,
    this.createdAt,
    this.priority = 'MEDIUM',
    this.dueDate,
    this.subtasks = const [],
    this.tags = const [],
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['id'],
      title: json['title'] ?? '',
      description: json['description'],
      completed: json['completed'] ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      priority: json['priority'] ?? 'MEDIUM',
      dueDate:
          json['dueDate'] != null ? DateTime.parse(json['dueDate']) : null,
      subtasks: json['subtasks'] != null
          ? (json['subtasks'] as List)
              .map((s) => SubTask.fromJson(s))
              .toList()
          : [],
      tags: json['tags'] != null
          ? List<String>.from(json['tags'])
          : [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'priority': priority,
      'dueDate': dueDate?.toIso8601String(),
      'subtasks': subtasks.map((s) => s.toJson()).toList(),
      'tags': tags,
    };
  }

  Task copyWith({
    int? id,
    String? title,
    String? description,
    bool? completed,
    DateTime? createdAt,
    String? priority,
    DateTime? dueDate,
    List<SubTask>? subtasks,
    List<String>? tags,
  }) {
    return Task(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      completed: completed ?? this.completed,
      createdAt: createdAt ?? this.createdAt,
      priority: priority ?? this.priority,
      dueDate: dueDate ?? this.dueDate,
      subtasks: subtasks ?? this.subtasks,
      tags: tags ?? this.tags,
    );
  }
}
